experimental/offline_attestation/shared/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Shared components used by both the client and server for offline attestation.
+
+use anyhow::anyhow;
+use serde::{Deserialize, Serialize};
+pub use tink_core::keyset::Handle;
+pub use tink_hybrid::init;
+
+const ENCRYPTION_CONTEXT: &[u8] = b"Offline Attestation v0.1";
+
+/// An encrypted request sent from the client to the server.
+#[derive(Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct EncryptedRequest {
+    /// The encrypted payload.
+    ///
+    /// Contains the raw output from Tink hybrid encryption.
+    pub ciphertext: Vec<u8>,
+
+    /// The public key that should be used for encrypting the response.
+    ///
+    /// Contains the binary-serialised form of the Tink keyset representing the public key.
+    pub response_public_key: Vec<u8>,
+}
+
+impl EncryptedRequest {
+    /// Deserialises the `response_publlic_key` as a keyset handle.
+    pub fn get_public_key_handle(&self) -> anyhow::Result<Handle> {
+        deserialize_public_key(&self.response_public_key)
+    }
+}
+
+/// An encrypted response sent from the server to the client.
+#[derive(Deserialize, Serialize)]
+pub struct EncryptedResponse {
+    /// The encrypted payload.
+    ///
+    /// Contains the raw output from Tink hybrid encryption.
+    pub ciphertext: Vec<u8>,
+}
+
+/// Information about the public key that should be used for encrypting requests.
+#[derive(Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PublicKeyInfo {
+    /// The public key that should be used for encrypting requests.
+    ///
+    /// Contains the binary-serialised form of the Tink keyset representing the public key.
+    pub request_public_key: Vec<u8>,
+
+    /// The binary-serialised attestation report providing evidence that the public key was
+    /// generated in the enclave and that the private key will not be leaked from the enclave.
+    ///
+    /// In a real implementation this will include a hash of the public key to bind the key to the
+    /// attestation.
+    pub attestation: Vec<u8>,
+}
+
+impl PublicKeyInfo {
+    /// Constructs a new `PublicKeyInfo` instance from the public key handle.
+    pub fn new(public_key_handle: &Handle) -> anyhow::Result<Self> {
+        // In a real implementation this would ask the enclave for an attestation report that binds
+        // the private key to the enclave. For now we just use and empty vector.
+        let attestation = Vec::new();
```

Comment:
I think it would still be nice to extract the "attestation" function, even if it's just implemented to return an empty vec for now. It would also start being useful when discussing interactive attestation in more detail.

Reply:
I added a placeholder AttestationReport struct with serialisation and deserialisation, and added a functions for generating the placeholder report.